src/tools/w3dviewer/datatreeview.cpp

```diff
@@ -639,33 +639,176 @@ void CDataTreeView::RestrictAnims(bool enable)
+                break;
+            case ASSET_TYPE_AGGREGATE:
+                item = this->m_categoryTreeItems[CATEGORY_AGGREGATE];
+                break;
+            case ASSET_TYPE_HLOD:
+                item = this->m_categoryTreeItems[CATEGORY_HLOD];
+                break;
+            case ASSET_TYPE_PARTICLEEMITTER:
+                item = this->m_categoryTreeItems[CATEGORY_EMITTER];
+                break;
+            case ASSET_TYPE_PRIMITIVE:
+                item = this->m_categoryTreeItems[CATEGORY_PRIMITIVES];
+                break;
+            case ASSET_TYPE_SOUND:
+                item = this->m_categoryTreeItems[CATEGORY_SOUNDS];
+                break;
+            default:
+                break;
+        }
+
+        HTREEITEM newitem = FindItemByName(item, oldname);
+
+        if (newitem != nullptr) {
+            GetTreeCtrl().SetItem(newitem, TVIF_TEXT, newname, 0, 0, 0, 0, 0);
+            AssetInfoClass *info = reinterpret_cast<AssetInfoClass *>(GetTreeCtrl().GetItemData(item));
+
+            if (info != nullptr) {
+                info->m_name = newname;
+            }
+        } else {
+            AddItem(newname, type, true);
+        }
+
+        SetRedraw();
+        InvalidateRect(nullptr, FALSE);
+        UpdateWindow();
+    }
 }
 
-void CDataTreeView::AddItem(const char *name, int type, bool select)
+void CDataTreeView::AddItem(const char *name, AssetType type, bool select)
 {
-    // do later
+    if (name != nullptr) {
+        SetRedraw(false);
+        HTREEITEM item = nullptr;
+        int image = 0;
+
+        switch (type) {
+            case ASSET_TYPE_MESH:
+                item = this->m_categoryTreeItems[CATEGORY_MESH];
+                image = this->m_imageListIDs[ICON_MESH];
+                break;
+            case ASSET_TYPE_AGGREGATE:
+                item = this->m_categoryTreeItems[CATEGORY_AGGREGATE];
+                image = this->m_imageListIDs[ICON_AGGREGATE];
+            case ASSET_TYPE_HLOD:
+                item = this->m_categoryTreeItems[CATEGORY_HLOD];
+                image = this->m_imageListIDs[ICON_HLOD];
+                break;
+            case ASSET_TYPE_PARTICLEEMITTER:
+                item = this->m_categoryTreeItems[CATEGORY_EMITTER];
+                image = this->m_imageListIDs[ICON_EMITTER];
+                break;
+            case ASSET_TYPE_PRIMITIVE:
+                item = this->m_categoryTreeItems[CATEGORY_PRIMITIVES];
+                image = this->m_imageListIDs[ICON_PRIMITIVE];
+                break;
+            case ASSET_TYPE_SOUND:
+                item = this->m_categoryTreeItems[CATEGORY_SOUNDS];
+                image = this->m_imageListIDs[ICON_SOUND];
+                break;
+            default:
+                break;
+        }
+
+        HTREEITEM newitem = nullptr;
+
+        for (HTREEITEM i = GetTreeCtrl().GetChildItem(item); i != nullptr; i = GetTreeCtrl().GetNextSiblingItem(i)) {
+            if (newitem) {

```

Comment:
This condition is before the other one in this `for` loop?

Reply:
Yes this is correct.